compiler/rustc_lint/src/lifetime_syntax.rs

```diff
@@ -71,7 +74,84 @@ declare_lint! {
     "detects when a lifetime uses different syntax between arguments and return values"
 }
 
-declare_lint_pass!(LifetimeSyntax => [MISMATCHED_LIFETIME_SYNTAXES]);
+declare_lint! {
+    /// The `hidden_lifetimes_in_input_paths` lint detects the use of
+    /// hidden lifetime parameters in types occurring as a function
+    /// argument.
+    ///
+    /// ### Example
+    ///
+    /// ```rust,compile_fail
+    /// #![deny(hidden_lifetimes_in_input_paths)]
+    ///
+    /// struct ContainsLifetime<'a>(&'a i32);
+    ///
+    /// fn foo(x: ContainsLifetime) {}
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Hidden lifetime parameters can make it difficult to see at a
+    /// glance that borrowing is occurring.
+    ///
+    /// This lint ensures that lifetime parameters are always
+    /// explicitly stated, even if it is the `'_` [placeholder
+    /// lifetime].
```

Comment:
Nit (wording): To make it easier to search for `'_`, maybe

> This lint ensures that lifetime parameters are always explicitly stated, even if it is the [placeholder lifetime `'_`][placeholder lifetime].

[placeholder lifetime]: https://doc.rust-lang.org/reference/lifetime-elision.html#lifetime-elision-in-functions

Actually hm, this is more so trying to express

> This lint ensures that lifetime parameters are explicit in types occuring as function arguments. Where lifetime elision may occur, the [placeholder lifetime `'_`][placeholder lifetime] may be used to make the presence of lifetime constraints explicit.